rts/Map/ReadMap.h

```diff
@@ -276,6 +281,8 @@ class CReadMap
 	float2 initHeightBounds; //< initial minimum- and maximum-height (before any deformations)
 	float2 currHeightBounds; //< current minimum- and maximum-height
 
+	std::map<const float, int> heightRefMap;
```

Comment:
There are no comments here, so I don't understand why you want to keep a count of every unique height value at all. What was wrong with tracking `currHeightBounds` the old way and why are refcounts the only possible fix? `UpdateHeightsRefMap` adds a ton of indirection (hidden via `std::map`, which is a terrible data structure in practice if you care about performance) and branching to a rather critical code path; this is **BAD** because the heightmap will be getting absolutely bombarded with updates during large-scale late-game play. Please reconsider your solution.

Reply:
I know it's less than ideal.
Here is the thing. The previous implementation was incorrect. It only worked fine if minimal height were getting decreased and maximal height - increased during the course of the game. It's not always the case:

* Terraform
* Waterlevel gadget (terraform internally too)
* Smoothing

@Floris spotted values were incorrect and it was also evident that costly bumpwater was sometimes get rendered although no piece of terrain was below 0.

With the above in mind the only solution I came up with is to track all heights in the map and use its properties to update min max on occasions.

This change has been live for couple of months at least and yes it costs something, but I have no other solution.

Reply:
Yeah I see the issue now, `currHeightBounds` only widens over time. I think an iterated update would be a good compromise here: on simframe `y`, scan heightmap row `y % num_rows`. Reset the working bounds to [+inf,-inf] when `y % n == 0`, keep a copy of the previous bounds while the iteration is ongoing.

Reply:
So you think processing a HM row every sim frame is better than tracking updates ad-hoc?

Reply:
Yes, especially when large areas are updated. `std::map` introduces cache misses that are avoided with (and hurt much more than) a simple linear sweep.

Reply:
Have a look https://github.com/spring/spring/pull/559/commits/985d21952633dc103acbb669e272aacfb1c50c85

Reply:
Looks good (just some minor nits: `const int idxBeg` etc), but on a 1024x1024 heightmap this will read ~35K samples per frame which might be pushing it. Can easily be tuned later though.

Reply:
I tried for_mt() approach on 20x20 map, so a chunk is around 54k items.
The results are pure disaster:

Even code like this:

```	for_mt(idxBeg, idxEnd, [this](const int idx) {
		const float h = (*heightMapSyncedPtr)[idx];
	});
```
consumes 7% of CPU time (I'm on 24 core machine). If I add atomics to get min/max it gets to 27%.
Do you know why for_mt is so costly?

Existing in-thread implementation consumes 0.2% compared.

Reply:
Moving a task into the pool hits the allocator, one or more worker threads have to be woken up (possibly all of them for `for_mt`), the queue needs to be managed, etc. It is meant more for longer-running background jobs where startup latency can be ignored.

Using atomics opens up another can of worms regarding cache coherency, letting each thread calculate bounds locally (i.e. over its own HM chunk) and merging them at the end is 99% likely to be faster.

Reply:
I was able to get slightly better times with integration and use of https://taskflow.github.io/taskflow/ParallelReduction.html#A2ParallelReduction , but for now it's probably not worth the bother.